### ISSUE TITLE
Choice of num_warps

### ISSUE THREAD

**Issue by Edenzzzz** [linkedin/Liger-Kernel#266]
I wonder why you use 32 (256 threads/block instance) here instead of deciding based on hidden size? Thanks.
https://github.com/linkedin/Liger-Kernel/blob/dd86cbd2092177681acf75643ded1b23a785a816/src/liger_kernel/ops/fused_linear_cross_entropy.py#L95

**Comment by DocShotgun** [comment]
Another thing to note with this, is that this hardcoded `num_warps` of 32 causes errors when trying to run on AMD Instinct accelerators, since they have a warp size of 64 rather than 32, and this ends up exceeding the max threads per block:

https://github.com/triton-lang/triton/issues/4128#issuecomment-2200944647
https://github.com/linkedin/Liger-Kernel/issues/231

**Comment by Edenzzzz** [comment]
Yeah I think this needs some more adaptive tuning. Though Triton generally seems to give good performance with brute-force parameters

**Comment by ByronHsu** [comment]
32 is the best tuned value i found for llama. did you observe better performance number if you use a different number?

**Comment by Edenzzzz** [comment]
Did you choose this based on hidden dim? Thanks.

**Comment by DocShotgun** [comment]
The hardcoded `num_warps` of 32 causes [this error](https://github.com/linkedin/Liger-Kernel/issues/231) on AMD Instinct accelerators, which I managed to avoid on my end by reducing it to 16 instead. I was subsequently able to full finetune Qwen2.5 72B at 32k ctx length and Mistral Large 123B at 16k ctx on 8xMI300X with all Liger Kernel optimizations enabled.

Perhaps it would be possible to automatically detect when an AMD Instinct accelerator is in use and adjust `num_warps` to a suitable value accordingly? However I'm not sure if this reduced value of 16 is optimal performance-wise on these GPUs.

**Comment by Edenzzzz** [comment]
It seems the the block size/vocab(32k or 64k at max) size will be much larger than whatever number of threads you have, so better set num_warps to the maximum available on your hardware?

**Comment by ByronHsu** [comment]
makes sense. feel free to send a pr for AMD!

**Comment by tjtanaa** [comment]
> It seems the the block size/vocab size(32k or 64k at max) will be much larger than any number of threads you have, so better set `num_warps` to the max available on your hardware?

I found that in ROCm/vllm and vllm fork, the restrict the `num_warps` to be either `4` or `8` in actual kernel, so maybe these numbers can be a reference to tune the `num_warps` on AMD.

https://github.com/search?q=repo%3AROCm%2Fvllm+num_warp&type=code&p=3
